blog/content/post/2019-05-15-building-bareflank.md

```diff
@@ -0,0 +1,345 @@
+being loaded and its signature - image must be signed with a key that is known
+to the previous stage.
+
+Most UEFI vendors implement only Microsoft keys for Secure Boot (*if it works
+with Windows, it works*), so other developers must ask Microsoft to sign their
+software (or ask the users to turn off Secure Boot). However,
+[Microsoft UEFI CA Signing policy](https://techcommunity.microsoft.com/t5/Windows-Hardware-Certification/Microsoft-UEFI-CA-Signing-policy-updates/ba-p/364828)
+puts restrictions on what can be signed by Microsoft. One of these restrictions
+is that `code submitted for UEFI signing must not be subject to GPLv3`, so
+GRUB2 can't be signed, even though it is aware of Secure Boot and continues the
+chain of trust.
+
+To deal with this problem, an additional signed stage must be added before
+GRUB2 (a bootloader for bootloaders). An example of such might be [PreLoader](https://blog.hansenpartnership.com/linux-foundation-secure-boot-system-released/)
+or [shim](https://github.com/rhboot/shim). The latter is [used by Ubuntu](https://wiki.ubuntu.com/UEFI/SecureBoot).
+
+This shim is included in live images, otherwise the users would be welcomed by
+a verification error printed by UEFI, with no pointers on how to start the
+system. The error message most definitely wouldn't ask them to turn the Secure
+Boot option off.
+
+All these problems result in an additional file in `\EFI\BOOT\` directory.
+There is an application signed by Microsoft in `bootx64.efi` (shim) and another
+one, `grubx64.efi` (GRUB2). GRUB is verified by the shim - shim was build with
+GRUB's certificate. With Secure Boot turned off `grubx64.efi` can be started
+directly.
+
+A similar approach should be possible with Bareflank, but for testing and
+fooling around it is easier to just turn the Secure Boot off.
+
+#### TL;DR
+
+Disable Secure Boot. Download and prepare a live USB drive of [Ubuntu](https://www.ubuntu.com/download/desktop).
+Optionally edit `boot/grub/grub.cfg` to enable serial output - add
+`console=ttyS0,115200` to the end of `linux /casper/vmlinuz ...` line, after
+the three dashes in the `Try Ubuntu without installing` section.
+
+By default, `bareflank.efi` tries to run `\EFI\BOOT\bootx64.efi` file. The same
+file is searched for in standard UEFI search for bootloaders. Because of that,
+it is difficult to start Bareflank and then the real bootloader automatically.
+There are basically 3 options for starting Bareflank. All files mentioned
+below (except `bareflank.efi`) are located in `\EFI\BOOT\` directory on the USB
+drive.
+
+###### 1st option
+
+Unless UEFI Shell can be forced to start despite a valid `bootx64.efi` file,
+the most automatic option is to modify the Bareflank source to start a different
+file. The line responsible for that is located in `load_start_vm()`, in
+`hypervisor/bfdriver/src/platform/efi/entry.c`. Change it to point to
+`grubx64.efi` to start GRUB directly and copy `bareflank.efi` to `bootx64.efi`.
```

Comment:
@krystian-hebel it would be great if you can provide ready to use patch in code section